File: Data Structures & Algorithms/depth-of-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int maxDepth(TreeNode* root) {
        if (root == nullptr) return 0;

        std::queue<TreeNode*> my_queue{};
        my_queue.push(root);
        int count = 0;
        while (!my_queue.empty()) {
            count++;
            int size = my_queue.size();
            for (int i = 0; i < size; i++) {
                TreeNode* node = my_queue.front();
                my_queue.pop();
                if (node->left != nullptr) my_queue.push(node->left);
                if (node->right != nullptr) my_queue.push(node->right);
            }
        }
        
        return count;
    }
};

File: Data Structures & Algorithms/valid-binary-search-tree/submission-1.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public bool IsValidBST(TreeNode root) {
        return Traversal(root, null, null);
    }

    private bool Traversal(TreeNode node, int? min, int? max) {
        if (node == null) return true;

        if (min.HasValue && node.val <= min) return false;
        if (max.HasValue && node.val >= max) return false;

        bool left = Traversal(node.left, min, node.val);
        if (!left) return false;
        bool right = Traversal(node.right, node.val, max);
        if (!right) return false;

        return true;
    }
}
